File: problems/Subarray Sum Equals K/subarraySum.cs
public class Solution {
    public int SubarraySum(int[] nums, int k) {
        var result = 0;
        var sum = 0;
        var store = new Dictionary<int, int> {{0, 1}};

        foreach (var item in nums) {
            sum += item;

            if (store.ContainsKey(sum - k)) {
                result += store[sum - k];
            }

            if (store.ContainsKey(sum)) {
                ++store[sum];
            } else {
                store.Add(sum, 1);
            }
        }

        return result;
    }
}

File: problems/Subarray Sum Equals K/subarraySum.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        return subarraySumOn1(nums, k);
        // return subarraySumOn2(nums, k);
    }

    private int subarraySumOn1(int[] nums, int k) {
        int result = 0;
        int sum = 0;
        HashMap<Integer, Integer> prefixSumToCount = new HashMap<>();

        prefixSumToCount.put(0, 1);

        for (int num : nums) {
            sum += num;

            if (prefixSumToCount.containsKey(sum - k)) {
                result += prefixSumToCount.get(sum - k);
            }

            prefixSumToCount.put(sum, 1 + prefixSumToCount.getOrDefault(sum, 0));
        }

        return result;
    }

    private int subarraySumOn2(int[] nums, int k) {
        int result = 0;
        int n = nums.length;
        int[] prefixSums = new int[1 + n];

        for (int idx = 0; n > idx; ++idx) {
            prefixSums[1 + idx] = nums[idx] + prefixSums[idx];
        }

        for (int i = 1; n >= i; ++i) {
            for (int j = 0; i > j; ++j) {
                if (k == prefixSums[i] - prefixSums[j]) {
                    ++result;
                }
            }
        }

        return result;
    }
}
